libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -897,6 +899,123 @@ std::string YulUtilFunctions::resizeDynamicArrayFunction(ArrayType const& _type)
 	});
 }
 
+string YulUtilFunctions::resizeDynamicByteArrayFunction(ArrayType const& _type)
+{
+	string functionName = "resize_array_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(array, newLen) {
+				if gt(newLen, <maxArrayLength>) {
+					<panic>()
+				}
+
+				let oldLen := <fetchLength>(array)
+				let offset := mod(newLen, 32)
+				let oldSlotCount := div(add(oldLen, 31), 32)
+				let newSlotCount := div(add(newLen, 31), 32)
+
+				if gt(newLen, oldLen) {
+					<increaseSize>(array, oldLen, newLen)
+				}
+
+				// Size was reduced, clear end of array
+				if lt(newLen, oldLen) {
+					switch lt(newSlotCount, 2)
+					case  0 {
+						let arrayDataStart := <dataPosition>(array)
+						let deleteStart := add(arrayDataStart, newSlotCount)
+						let deleteEnd := add(arrayDataStart, oldSlotCount)
+
+						// we have to partially clear last slot that is still used, so decreasing start by one
+						if gt(offset, 0) { <partialClearStorageSlot>(sub(deleteStart, 1), offset) }
+
+						<clearStorageRange>(deleteStart, deleteEnd)
+
+						sstore(array, or(<shl>(1, newLen), 1))
+					}
+					default {
+						<decreaseSizeToPacked>(array, oldLen, newLen)
+					}
+				}
+			})")
+			("functionName", functionName)
+			("panic", panicFunction())
+			("fetchLength", arrayLengthFunction(_type))
+			("dataPosition", arrayDataAreaFunction(_type))
+			("clearStorageRange", clearStorageRangeFunction(*_type.baseType()))
+			("maxArrayLength", (u256(1) << 64).str())
+			("partialClearStorageSlot", partialClearStorageSlotFunction())
+			("decreaseSizeToPacked", decreaseByteArraySizeToPackedFunction(_type))
+			("increaseSize", increaseByteArraySizeFunction(_type))
+			("shl", shiftLeftFunctionDynamic())
+			.render();
+	});
+}
+
+string YulUtilFunctions::decreaseByteArraySizeToPackedFunction(ArrayType const& _type)
+{
+	string functionName = "byte_array_decrease_size_to_packed_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(array, oldLen, newLen) {
+				let oldSlotCount := div(add(oldLen, 31), 32)
+
+				let arrayDataStart := array
+				if gt(oldSlotCount, 1) {
```

Comment:
I'm a bit confused by this. Does this work properly for length == 32? I would find it safer to work with the exact length here.

Reply:
Actually it wasn't working correctly. Should be better now when I fixed it.